topics/dapp-wallet-msg.md

```diff
@@ -0,0 +1,110 @@
+# Messages
+
+## Permission
+
+A `permission_request` message is sent in order to establish a connection:
+1. The relying party sends `permission_request` to the wallet. The message includes:
+    - basic information about the relying party (`appMetadata`),
+    - a list of networks on which the relying party plans to operate (`networks`),
+    - a list of permission scopes it requires (`scopes`).
+2. The wallet receives the message, presents the details of the to-be-established connection to the user and asks the user to select the account that will be paired in response.
+    - If the user approves the request, the wallet sends `permission_response` back to the relying party. The message includes:
+        - basic information about the wallet (`appMetadata`),
+        - a list of networks on which the user has agreed the relying part can operate, should be a subset of the equivalent from `permission_request` (`networks`),
+        - a list of permission scopes the user has agreed the relying party can be granted, should be a subset of the equivalent from `permission_request` (`scopes`)
+        - the account details (princial or public key (*TBD what is required*) and optional subaccount) of the user selected account (`account`)
+    - If the user rejects the request, the wallet sends `error` back to the relying party with appropriate explanation.
+
+```
+{ // Request
+    type: "permission_request"
+    appMetadata: /* dApp */ {
```

Comment:
Yes, both sides create key pairs and exchange the public keys. The public key is the true source of identity which the user can use to verify that the request is coming from a dApp they trust. 

Normally, this public key is part of the message, but I omitted it deliberately for the first draft to focus solely on the purely IC-related content to make sure that we go in the right direction. Since I got my confirmation, I guess I can clutter it now with the technicalities 😅

Reply:
Unfortunately, there's nothing stopping someone from creating a malicious copy of a popular dApp and try to completely pose as the original (UI wise, not only via internal channels of dApp and wallet communication). To prevent that, the wallet would have to know the dApp and its public key beforehand, meaning that it would have to use some kind of an allow list.

Reply:
One approach could also be to incorporate information provided by the browser. I.e. if the communication with the wallet is via `windowPostMessage` then the `origin` of the message is set by the browser and cannot be faked by a malicous dapp.
But this scheme only works over some channels unfortunately and is not a generic solution.